docs/SOFTWARE_DESIGN.md

```diff
@@ -0,0 +1,124 @@
+
+Unit testing will be performed on the C++ game engine code using [`utest.h`](https://github.com/sheredom/utest.h).
+
+The browser user interface will be tested with [Jest](https://github.com/facebook/jest) and [Puppeteer](https://github.com/puppeteer/puppeteer/).
+
+The turn-based nature of puzzle games and the HTML user interface makes it very easy to test our code, as opposed to a live action game on a screen where it'd be very difficult to programmatically provide input and assert on game state. Real-time graphics-based games can be math-heavy and asset-heavy.
+
+There is a very clear separation between front- and back-ends, so we can delegate disjoint tasks and responsibilities. The back-end game logic can be written in C++ without any knowledge of JS (and could ideally be modular enough to couple with a command-line front-end that might be useful for development). Greg will mostly be responsible for linking the JS and C++ components together.
+
+```
+           BACKEND (game engine)
+
+       .-------.
+       | uTest |
+       `-------`
+           |
+           v        [emscripten compilation]
+.-----------------.          |
+| C++ game engine |----------+--+
+`-----------------`             |
+                                |
+................................|.............
+                                |
+.------.   FRONTEND (browser)   |
+| User |                        |
+`------`                        |
+   |                            |
+[click events]    [API calls]   |
+   |                   |        v
+   |   .----------.~~~~+~~~>.------.
+   +-->| HTML5 UI |         | WASM |<--compiled backend engine
+       `----------`<~~~+~~~~`------`
+            ^          |
+            |     [game state]
+            |
+            |
+.----------------------.
+| Jest/Puppeteer tests |
+`----------------------`
+```
+
+Greg has built a simple proof-of-concept that illustrates the concept/workflow and relevant technologies interoperating: <https://github.com/ggorlen/wasm-click-counter>.
+
+### Game engine back-end
+
+The Sokoban game engine will consist of a top-level C++ class, `Sokoban`, which will expose the following interface:
+
+```
+// one of the possible objects that can exist on a cell
+enum Cell {
+   ' ': empty space
+   '#': impassable wall
+   '.': goal cell
+   '$': box
+   '*': box on top of goal cell
+   '@': player
+   '+': player on top of goal cell
+}
+
+// one of the possible movement directions for the player
+enum Direction {UP, DOWN, LEFT, RIGHT}
+
+// attempt a player move in direction (up/down/left/right) and return true/false to indicate if it was successful
+bool move(Direction direction)
+
+// switch to level level_number if possible, throw otherwise
+void change_level(int level_number)
+
+// reset the current level to its original state (could be replaced by change_level by passing in the current level number)
+void reset()
+
+// undo the last move
+void undo()
```

Comment:
For this function, we need a string variable to append every move the user make and reverse the direction so we can utilize this function, right? So if the user entered UUDDLR. We pop R and convert it to L. Maybe we need a reverse(String input) function.

Reply:
Yeah, that's the general idea. I'd think something like `reverse` would be an internal or utility function if we need it. The goal here is to sketch out the public interface we want to expose to the API wrapper and ultimately to JS.